File: 15-3sum/3sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        result = list()
        nums.sort()

        for index, number in enumerate(nums):
            if index > 0 and number == nums[index - 1]:
                continue
            
            left, right = index + 1, len(nums) - 1
            while left < right:
                sum_of_three = number + nums[left] + nums[right]

                if sum_of_three < 0:
                    left += 1
                elif sum_of_three == 0:
                    result.append([nums[index], nums[left], nums[right]])
                    left += 1
                    while nums[left] == nums[left - 1] and left < right:
                        left += 1 
                else:
                    right -= 1

        return result

File: 15-3sum/3sum.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        vector<vector<int>> result;
        int n = nums.size();
        sort(nums.begin(), nums.end());

        unordered_map<int, int> lastIndex;
        for (int i = 0; i < n; ++i) {
            lastIndex[nums[i]] = i;
        }

        for (int i = 0; i < n - 2; ++i) {
            if (i > 0 && nums[i] == nums[i - 1]) continue;

            for (int j = i + 1; j < n - 1; ++j) {
                if (j > i + 1 && nums[j] == nums[j - 1]) continue;

                int complement = -(nums[i] + nums[j]);
                
                if (lastIndex.count(complement) && lastIndex[complement] > j) {
                    result.push_back({nums[i], nums[j], complement});
                }
            }
        }

        return result;
    }
};